pkg/sessions/redis/lock.go

```diff
@@ -0,0 +1,84 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/bsm/redislock"
+	"github.com/go-redis/redis/v8"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+)
+
+var LockSuffix = "lock"
+
+type Lock struct {
+	client redis.Cmdable
+	locker *redislock.Client
+	lock   *redislock.Lock
+	key    string
+}
+
+// NewLock instantiate a new lock instance. This will not yet apply a lock on Redis side.
+// For that you have to call Obtain(ctx context.Context, expiration time.Duration)
+func NewLock(client redis.Cmdable, key string) sessions.Lock {
+	return &Lock{
+		client: client,
+		locker: redislock.New(client),
+		key:    key,
+	}
+}
+
+// Obtain obtains a distributed lock on Redis for the configured key.
+func (l *Lock) Obtain(ctx context.Context, expiration time.Duration) error {
+	lock, err := l.locker.Obtain(ctx, l.lockKey(), expiration, nil)
+	if errors.Is(err, redislock.ErrNotObtained) {
+		return sessions.ErrLockNotObtained
+	}
+	if err != nil {
+		return err
+	}
+	l.lock = lock
+	return nil
+}
+
+// Refresh refreshes an already existing lock.
+func (l *Lock) Refresh(ctx context.Context, expiration time.Duration) error {
```

Comment:
I'm curious on Refresh & Release:

If I recall, the underlying `redislock` library has lock protection be having a cryptographic nonce in the value. The locker also has that value and any modification to the lock needs to validate that the caller knows the cryptographic nonce (protects against a non-holder of the nonce from trampling by accessing the same lock key).

What is the rationale for messing with the lock keys in Redis directly without using the `locker` that gives these protections?

`Release` in particular I wonder if there's risk of trampling another lock holder's lock?

Reply:
https://github.com/bsm/redislock/blob/162618b021216846fc29f90b1c58f2620131714b/redislock.go#L133

Reply:
Thats' a good point. I did it to actually enable this specific case. But you are right, it should not be like that. I will update it by using the lock instance for refreshing and releasing.